src/main/kotlin/com/github/gradle/android/i18n/import/XlsImporter.kt

```diff
@@ -1,64 +1,217 @@
-                        .asSequence()
-                        .filter { it.columnIndex > 0 }
-                        .filter { it.stringCellValue.isNotBlank() }
-                        .forEach { cell ->
-                            val locale = cell.stringCellValue
-                            handler.addLocale(locale)
-                            localizedColumns[cell.columnIndex] = locale
-                        }
-            } else if (row.rowNum > 0) {
-                // Reading other data row.
-                val key = row.getCell(0)?.stringCellValue
-
-                val translations = localizedColumns
-                        .map { Pair(it.value, row.getCell(it.key)?.stringCellValue) }
-                        .toMap()
-
-                handler.addTranslations(key, translations)
-            }
+class XlsImporter(private val project: Project) : AbstractImporter(project) {
+
+    override fun generate(inputStream: InputStream, config: ImportConfig) {
+
+        val workbook = WorkbookFactory.create(inputStream) // TODO .use
+
+        val projectData = if (project.isSingleModule()) {
+            workbook.toSingleModuleProjectData(config)
+        } else {
+            workbook.toMultiModuleProjectData(config)
         }
 
-        handler.writeOutput()
+        val stringResourcesByPath = projectData.toStringResourcesByPath(project.projectDir, config)
+        stringResourcesByPath.write()
     }
+}
 
-    private fun readInput(inputStream: InputStream, config: ImportConfig, consumer: (Row, Boolean) -> Unit) {
-        var readLocales = true
-        WorkbookFactory.create(inputStream).forEachIndexed { index, sheet ->
-            when {
-                config.allSheets -> if (config.sheetNameRegex.matches(sheet.sheetName)) {
-                    // All sheets matching name regex.
-                    sheet.forEach { consumer(it, readLocales) }
-                    readLocales = false
-                }
+private fun Project.isSingleModule(): Boolean {
+    return this.childProjects.size <= 1
+}
+
+private fun Workbook.toMultiModuleProjectData(config: ImportConfig): ProjectData {
+    val moduleDataList = this.map { sheet: Sheet ->
+        val stringDataByLocale = sheet.toList().toStringDataByLocale()
+        val translationDataList = stringDataByLocale.keys.map { locale ->
+            TranslationData(locale, stringDataByLocale[locale] as List<StringData>)
+        }
+        ModuleData(sheet.sheetName, translationDataList)
+    }
+    return ProjectData(moduleDataList)
+}
+
+private fun Workbook.toSingleModuleProjectData(config: ImportConfig): ProjectData {
+    val rows = readWorkbookRows(config, this)
+    val stringDataByLocale = rows.toStringDataByLocale()
+    val translationDataList = stringDataByLocale.keys.map { locale ->
+        TranslationData(locale, stringDataByLocale[locale] as List<StringData>)
+    }
+    val moduleData = ModuleData("android-gradle-i18n", translationDataList)
```

Comment:
Why the module name is given ? it shouldn't be the name of the first sheet or app or project name ?

Reply:
This is to be consistent with export. When exporting resources of a single module project, the Excel file only has one sheet whose name is hardcoded ⇒ to do: use the same constant for this hardcoded name.